### ISSUE TITLE
How are source maps intended to work?

### ISSUE THREAD

**Issue by JonathanHolvey** [leafo/scssphp#550]
What is the expected output when using the option `$scss->setSourceMap(Compiler::SOURCE_MAP_INLINE)`?

Will the example server require modification to serve source map files in the same way it does compiled SCSS?

**Comment by dleffler** [comment]
Here's an example of code required to generate source maps, either inline or file based. https://github.com/leafo/scssphp/wiki/Source-Maps 

I'm not sure the level of your understanding, but source maps are used to help debug (scss) files since most browsers (chrome, firefox, etc...) can use them in association with their Developer Tools to point to the specific source file line instead of simply pointing to the compiled output stylesheet. E.g., Twitter Bootstrap 4 is composed of MANY .scss files, but creates only one .css file.

**Comment by chrisblakley** [comment]
I didn't realize there was a wiki entry for this- it was helpful, thank you.

**Comment by mahagr** [comment]
I find source map generation very limited -- you have control to write CSS files manually, but not .map files. IMHO it would have been much better design to do something like this:

```
    $scss = new Compiler();
    $scss->setSourceMap(Compiler::SOURCE_MAP_FILE);
    $css = $scss->compile($scssSource);
    $map = $scss->getSourceMapArray();  // You can also use getSourceMapJson() to get JSON encoded version
    // Save both files manually by the application, maybe allowing some post-processing on them...
```

I'm using streams (gantry-theme://) to locate the files and the streams need to be resolved in order to make source maps to work. But even without this, I don't think that SCSS compiler should save the map files, at least not by default.

**Comment by JonathanHolvey** [comment]
A source map implementation in `example/Server.php` would be really useful as well. It would be great if the source map files could be served dynamically from the SCSS source folder, without having to exist in the file system.

**Comment by mahagr** [comment]
FYI, this is my workaround for the map issues I was having:

```php
    $scss->setSourceMap(Compiler::SOURCE_MAP_INLINE);
    $scss->setSourceMapOptions([
        'sourceMapBasepath' => '/',
        'sourceRoot'        => '/',
    ]);
    $css = $scss->compile($scssSource);

    // Extract map from css and save it as separate file.
    if ($pos = strrpos($css, '/*# sourceMappingURL=')) {
        $map = json_decode(urldecode(substr($css, $pos + 43, -3)), true);
        foreach ($map['sources'] as &$source) {
            // TODO: Modify source files here...
        }
        unset($source);

        // TODO: Save map file here...

        $css = substr($css, 0, $pos) . '/*# sourceMappingURL=' . basename($out) . '.map */';
    }
```

**Comment by mahagr** [comment]
It could be the best if you could pass the source map class to the compiler instead of using a const to set it up. In that way, you could write your own logic to handle the maps... :)

**Comment by dleffler** [comment]
So, if I understand you correctly, you would add a 4th option to the types of Source Maps? (no map, inline map, file map, and custom class)?

In your thinking would this (always) be a class inherited (extends) from SourceMapGenerator and (always) be passed as an already instantiated object? perhaps more like:
```php
    $scss->setSourceMap(new MyOwnSourceMapGenerator($source_map_options_array);
```

**Comment by dleffler** [comment]
Therefore, if the answer to the above questions is (always) YES, the fix would be to change the code in Compiler.php around line #210 from
```php
    if ($this->sourceMap && $this->sourceMap !== self::SOURCE_MAP_NONE) {
        $sourceMapGenerator = new SourceMapGenerator($this->sourceMapOptions);
    }
```
to
```php
    if ($this->sourceMap) {
        if (is_object($this->sourceMap) && is_a($this->sourceMap, 'SourceMapGenerator')) {
            $sourceMapGenerator = $this->sourceMap;
        } elseif ($this->sourceMap !== self::SOURCE_MAP_NONE) {
            $sourceMapGenerator = new SourceMapGenerator($this->sourceMapOptions);
        }
    }
```
I'll try to test this out and perhaps create a pull request

**Comment by dleffler** [comment]
I've got a pull request up, #553

**Comment by robocoder** [comment]
Merged and tagged v0.7.5
